--- a/kernel-open/nvidia/os-interface.c
+++ b/kernel-open/nvidia/os-interface.c
@@ -3028,7 +3028,12 @@
 #if defined(NV_DMEM_CGROUP_PRESENT)
 void* NV_API_CALL os_dmem_cgroup_register_region(const char *name, NvU64 size, NvU64 precharge, void **prechargePool)
 {
-    void *region = dmem_cgroup_register_region(size, name);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(7, 3, 0)
+    const struct dmem_cgroup_init init = { .size = size };
+    void *region = dmem_cgroup_register_region(&init, "%s", name);
+#else
+    void *region = dmem_cgroup_register_region(size, "%s", name);
+#endif
     if (IS_ERR(region))
     {
         return NULL;
